man/kernel-install.xml

```diff
@@ -224,7 +223,7 @@
       (EFI System Partition) are mounted, and also conceptually referred to as <varname>$BOOT</varname>. Can
       be overridden by setting <varname>$BOOT_ROOT</varname> (see below).</para>
 
-      <para><varname>$KERNEL_INSTALL_LAYOUT=bls|uki|other|...</varname> is set for the plugins to specify the
```

Comment:
Not entirely sure that's the best solution, since this makes this harder to configure. I think a solution like mentioned in [this comment](https://github.com/systemd/systemd/issues/25822#issuecomment-1364105685) would be preferable, since having UKIs with type 1 entries might become a usecase down the road.

Reply:
If we want make this configurable I don't think this should be `layout=...` but a new, separate switch.

Reply:
I'm not sure it makes sense to add another option, when it leads back to `KERNEL_INSTALL_LAYOUT` having a single valid option, which historically meant BLS type 1.

Reply:
UKI and non-UKI kernels are different things, having a single option for both doesn't make sense to me.  Specifically the problem with `layout=uki` is that installing a non-UKI kernel in `/EFI/Linux` will simply not work in most cases because the initrd is missing then ...

Reply:
That is true, but if you choose to do UKIs and you are installing something else there, than there's already an issue with whatever tool generates your UKIs, no? Maybe I'm misunderstanding something.

Reply:
That may work on your local machine.  Doing the switch to UKIs at distro-level is not going to work that way.  We will see UKIs and non-UKIs living side-by-side for quite a while, maybe forever.  Fedora will ship kernel rpm packages for both UKI and non-UKI kernels.  So having UKI and non-UKI kernels installed at the same time can easily happen and there are valid reasons to do that.  For testing and development.  Or just temporary when switching a machine from one kind of kernels to the other (install new, reboot, uninstall old).

More details: https://fedoraproject.org/wiki/Changes/Unified_Kernel_Support_Phase_1

Reply:
There is `layout=bls` which will effectively activate `90-loaderentry`, which installs the kernel according to BLS to `$BOOT/$MACHINE_ID/$KERNELVER/linux` and creates a BLS type1 config snippet for it.

There is `layout=other` (or layout being unset) where the systemd-shipped scripts don't do anything.  In case of Fedora a script dropped by the grub rpms runs in that case and installs the kernel and updates the config.  It's actually a BLS type1 entry too (fedora grub is patched to understand that), but kernel + initrd are placed elsewhere (`$BOOT/vmlinuz-$KERNELVER`).  This is Fedoras default behavior today.

So there are already two possible options for `layout=`.

For UKIs the default should be to install them to `/EFI/Linux`, i.e. BLS type 2.  But flipping the default for non-UKI kernels to BLS type 2 doesn't make sense at all.  Installing UKIs as type 1 might make sense though if you want add command line options, so we should probably allow that.  So we could have:

```
layout={bls,other}
uki_layout={uki,bls,other}
```

Oh, and expecting users and developers work with different config files and/or setting `KERNEL_INSTALL_*` is not going to fly in practice.  People will get it wrong all the time then.

Reply:
We *will* have both types of images installed on the same system, so the tools should support flexible operation. The two scenarios that are "obvious":
1. traditional BLS type#1 installations with a boot loader entry and linux and initrd
2. pre-built UKI delivered via a package

but there are some plans to also support:

3. a mode where you reference a UKI and create a BLS type#1 entry for it with some override of options
4. building a UKI on demand

3. is useful when you want to override options and don't want to pay the price (or maybe even can't) of having two sets of kernel+initrd installed. 4. is useful when you want to use boot measurement or attestation but the distro doesn't give you the right kernel or you are building something locally.

I'd like to work backwards from those scenarios and figure out what the interface should be. But frankly, I don't quite see how each case would work yet.